tools/SourceKit/include/SourceKit/Core/LangSupport.h

```diff
@@ -220,6 +220,60 @@ struct RawCharSourceRange {
   unsigned Length;
 };
 
+enum class MacroRole : uint8_t {
+  // This should align with 'swift::MacroRole'.
+  Expression = 0x01,
+  Declaration = 0x02,
+  Accessor = 0x04,
+  MemberAttribute = 0x08,
+  Member = 0x10,
+  Peer = 0x20,
+  Conformance = 0x40,
+  CodeItem = 0x80,
+};
+using MacroRoles = swift::OptionSet<MacroRole>;
+
+struct MacroExpansionInfo {
+  // See swift::ExternalMacroReference.
+  struct ExternalMacroReference {
+    std::string moduleName;
+    std::string typeName;
+
+    ExternalMacroReference(StringRef moduleName, StringRef typeName)
+        : moduleName(moduleName), typeName(typeName){};
+  };
+  // See swift::ExpandedMacroDefinition.
+  struct ExpandedMacroDefinition {
+    struct Replacement {
+      RawCharSourceRange range;
+      unsigned parameterIndex;
```

Comment:
Your understanding is correct. I will add some comment about it.
As for variadics, I am not sure tbh. This just follows the compiler's implementation.

Reply:
Also, since "expanded" is not fully supported for attached macro, I don't think we can fully support it.
That's why I didn't mention it in the description.

Reply:
Looks like there is some issue with variadic macros defined in terms of other macros in general. Filed rdar://110419352